File: listas/lista_4/src/trapezoidal.cpp
#include <vector>
#include "xtensor/xarray.hpp"
#include "xtensor/xmath.hpp"
#include "xtensor/xio.hpp"
#include "xtensor/xview.hpp"
#include "xtensor/xadapt.hpp"
#include "xtensor/xindex_view.hpp"

#include "trapezoidal.hpp"

xt::xarray<double> Apply(double (*f)(double), xt::xarray<double> xData)
{
    std::vector<double> fy;

    for(double x: xData)
    {
        fy.push_back(f(x));
    }
    return xt::adapt(fy, xData.shape());
}

double NumericalIntegration::TrapezoidalRule(double (*f)(double), double integralStart, double integralEnd)
{
    double steps = (integralEnd - integralStart);
    xt::xarray<double> yValues = Apply(f, xt::arange(integralStart, integralEnd + steps, steps));
    xt::xarray<double> result = xt::sum(yValues) * ((integralEnd - integralStart) / 2.0);

    return result.data()[0];
}

double NumericalIntegration::TrapezoidalRuleRepeated(double (*f)(double), double integralStart, double integralEnd, double parts)
{
    double h = (integralEnd - integralStart) / parts;
    xt::xarray<double> rangeValues = xt::arange(integralStart, integralEnd + h, h);
    xt::xarray<double> yValues = Apply(f, xt::filter(rangeValues, rangeValues <= integralEnd));
    
    xt::xarray<double> trapezoidalBetweenInterval = 
                            xt::view(yValues, xt::drop(0, yValues.size()), xt::all());
    xt::xarray<double> trapezoidalArea = yValues(0) + 
                            ( 2.0 * xt::sum(trapezoidalBetweenInterval) ) + yValues(yValues.size() - 1);
    
    return trapezoidalArea.data()[0] * (h / 2.0);
}

double NumericalIntegration::TrapezoidalRulePoints(xt::xarray<double> yValues, double integralStart, double integralEnd, double parts)
{    
    double h = (integralEnd - integralStart) / parts;
    
    xt::xarray<double> trapezoidalBetweenInterval = 
                            xt::view(yValues, xt::drop(0, yValues.size()), xt::all());   
    xt::xarray<double> trapezoidalArea = yValues(0) + 
                            ( 2.0 * xt::sum(trapezoidalBetweenInterval) ) + yValues(yValues.size() - 1);
    
    return trapezoidalArea.data()[0] * (h / 2.0);
}

File: listas/lista_4/src/trapezoidal.py
# -*- coding: utf-8 -*-
import numpy as np
import scipy.interpolate
import matplotlib.pyplot as plt

def trapezoidal_rule_fnc(fnc, integral_start, integral_end, parts):
    h = (integral_end - integral_start) / parts
    interval = np.arange(integral_start, integral_end + h, h)
    
    y_values = fnc(interval[interval <= integral_end])
    result = y_values[0] + (2 * np.sum(y_values[1:-1])) + y_values[-1]
    
    return result * (h / 2)

x = list(range(0, 22, 2))
fx = [0, 1.8, 2., 4., 4., 6., 4., 3.6, 3.4, 2.8, 0]

fx_interp = scipy.interpolate.interp1d(x, fx)

results = [trapezoidal_rule_fnc(fx_interp, 0, 20, i) for i in range(10, 50)]
